feat: (bb) 128-bit challenges

zac-williamson opened the pull request:
This PR modifies our Transcript class to achieve the following:

every time a hash function is used to generate challenges, the output is split into two 128-bit field elements to generate 2 challenges per hash

This change gives us the following benefits:

1. the amount of hashing required to fold/verifier proofs is reduced
2. where challenges map to Verifier scalar multiplications, those scalar muls are now half-width and can be more efficiently evaluated (requires additional code to support)

Closes https://github.com/AztecProtocol/barretenberg/issues/741.

zac-williamson committed:
transcript splits every hash into two 128/126 bit challenges

zac-williamson committed:
compiler fixes

lucasxia01 commented on barretenberg/cpp/src/barretenberg/stdlib/transcript/transcript.hpp:
```diff
@@ -19,7 +19,23 @@ template <typename Builder> struct StdlibTranscriptParams {
+     * @param challenge
+     * @return std::array<Fr, 2>
+     */
+    static inline std::array<Fr, 2> split_challenge(const Fr& challenge)
```
not too familiar with cycle_scalar... why are we using it instead of normal field_ts?

zac-williamson replied:
cycle_scalar has an existing method to split a stdlib::field_t element into two half-width scalars. It's nontrivial to do efficiently so I wanted to reuse existing code

lucasxia01 replied:
it's a little unfortunate we have to bring in cycle group (which shouldn't have anything to do with this) but alright. I guess I don't have a better solution beyond copy paste... or some bigger refactoring

lucasxia01 commented on barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/circuit_simulator.hpp:
```diff
@@ -38,6 +40,7 @@ namespace bb {
 class CircuitSimulatorBN254 {
   public:
     using FF = bb::fr;
+    using EmbeddedCurve = std::conditional_t<std::same_as<FF, bb::g1::coordinate_field>, curve::BN254, curve::Grumpkin>;
```
where does this come in?

zac-williamson replied:
we were not explicitly instantiating `cycle_scalar` using the CircuitSimulatorBN254 class. In order to simulate the new stdlib::transcript class (which now has `cycle_scalar` as a dependency) I needed to explicitly instantiate it.

`cycle_scalar` expects the Builder template parameter to have an `EmbeddedCurve` type. This mirrors the code in the other circuit builders.

lucasxia01 commented on barretenberg/cpp/src/barretenberg/transcript/transcript.hpp:
```diff
@@ -71,6 +71,24 @@ struct NativeTranscriptParams {
+     */
+    static inline std::array<Fr, 2> split_challenge(const Fr& challenge)
+    {
+        static constexpr size_t LO_BITS = plookup::FixedBaseParams::BITS_PER_LO_SCALAR;
```
weird to use plookup::FixedBaseParams::BITS_PER_LO_SCALAR, maybe we can move this constant somewhere else/rename it?

lucasxia01 replied:
also, shouldn't we try to generate 2 127 bit challenges as opposed to a 128 and a 126 bit challenge?

zac-williamson replied:
yeah this could do with a renaming

zac-williamson replied:
> also, shouldn't we try to generate 2 127 bit challenges as opposed to a 128 and a 126 bit challenge?

the extra bit of security isn't significant as the curve we use has 100-110 bits of security. I felt having a simple "cut this into 128-bit chunks" heuristic would be simpler to reason about.

also...splitting into 128/126 bits is 1 gate cheaper. `create_range_constraint` can cut 126 bits into 9 14-bit slices perfectly, which requires 3 gates to accumulate. 127/128 bits require 10 slices which needs 4 gates.

zac-williamson replied:
> weird to use plookup::FixedBaseParams::BITS_PER_LO_SCALAR, maybe we can move this constant somewhere else/rename it?

I moved the definition into `field::Params` and its naming reflects why it's 128 (at least for `cycle_scalar`, it's to do with how large the scalar value can be when splitting a field element into two half-width slices according to the endormorphism trick)

lucasxia01 commented on barretenberg/cpp/src/barretenberg/transcript/transcript.hpp:
```diff
@@ -128,7 +146,7 @@ template <typename TranscriptParams> class BaseTranscript {
      * @return std::array<Fr, HASH_OUTPUT_SIZE>
      */
-    [[nodiscard]] Fr get_next_challenge_buffer()
+    [[nodiscard]] std::array<Fr, 2> get_next_duplex_challenge_buffer(size_t num_challenges)
```
please update this so that it checks num_challenges is 1 or 2, or at least update some comment

zac-williamson replied:
done

lucasxia01 reviewed (commented):
Looks mostly fine. Left some questions/comments. Also, does this actually provide much benefit vs just truncating challenges to 128 bits? (Edit: ah, maybe we try to generate all the alphas in a row so it cuts down hashing by almost 50% there..., but otherwise maybe not too much impact elsewhere...)

zac-williamson commented:
> Looks mostly fine. Left some questions/comments. Also, does this actually provide much benefit vs just truncating challenges to 128 bits? (Edit: ah, maybe we try to generate all the alphas in a row so it cuts down hashing by almost 50% there..., but otherwise maybe not too much impact elsewhere...)

yeah it's mostly the alpha challenges. There are over 40 of them IIRC, so removing 20 hashes is nontrivial. There are a few other hashes it cuts out IIRC we estimated the saving could be up to 3k constraints

zac-williamson committed:
bug fix (stdlib transcript and native transcript now split scalars the same way)

moved the magic number definition for the number of bits in the lo challenge scalar to `field::Params`

zac-williamson committed:
bug fix

ledwards2225 committed:
update acir test fixtures

lucasxia01 committed:
regenerated input to verify_honk_proof test program due to proof change

zac-williamson committed:
oink alpha challenges are now generated simultaneously instead of one alpha challenge per round

zac-williamson committed:
updated Honk Solidity verifier

zac-williamson committed:
regenerated cached proofs

zac-williamson committed:
regenerated keys/proofs

zac-williamson committed:
updated honk_contract

zac-williamson committed:
wip

zac-williamson committed:
wip

vezenovm committed:
update generateTranscript call in honk_contract.hpp

zac-williamson committed:
updated honk_contract

zac-williamson committed:
Merge branch 'zw/split-challenges' of github.com:AztecProtocol/aztec-packages into zw/split-challenges

zac-williamson committed:
reverted script

vezenovm committed:
fix generateTranscript sig

vezenovm committed:
fix generateEtaChallenge signature

vezenovm committed:
removed accidentally committed yarn lock

zac-williamson committed:
removed code duplication

lucasxia01 committed:
removed commented out code, added issue

lucasxia01 commented on barretenberg/cpp/src/barretenberg/dsl/acir_proofs/honk_contract.hpp:
```diff
@@ -292,34 +292,45 @@ struct Transcript {
+        second = FrLib.fromBytes32(bytes32(hi));
+    }
+
+    function generateEtaChallenge(Honk.Proof memory proof, bytes32[] calldata publicInputs, uint256 publicInputsSize)
```
publicInputsSize doesn't seem needed here as an input

lucasxia01 replied:
but maybe this was needed to maintain the same interface as Transcript.sol? unsure, didn't want to test it so I'll leave it

lucasxia01 committed:
Merge remote-tracking branch 'origin/master' into zw/split-challenges

lucasxia01 reviewed (approved):
Don't like a couple things - using cycle_scalar in the transcript, and the nasty C++ code used to turn one argument that's array of strings to multiple arguments of the strings.

Solidity changes seem fine in general. Overall, good stuff. Hopefully, we can document some of the effects of this PR on gates somewhere as well.
